File: Python/get-equal-substrings-within-budget.py
# Time:  O(n)
# Space: O(1)

class Solution(object):
    def equalSubstring(self, s, t, maxCost):
        """
        :type s: str
        :type t: str
        :type maxCost: int
        :rtype: int
        """
        left = 0
        for right in xrange(len(s)):
            maxCost -= abs(ord(s[right])-ord(t[right]))
            if maxCost < 0:
                maxCost += abs(ord(s[left])-ord(t[left]))
                left += 1
        return (right+1)-left

File: C++/get-equal-substrings-within-budget.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    int equalSubstring(string s, string t, int maxCost) {
        int left = 0, right = 0;
        for (; right < s.length(); ++right) {
            maxCost -= abs(s[right] - t[right]);
            if (maxCost < 0) {
                maxCost += abs(s[left] - t[left]);
                ++left;
            }
        }
        return right - left;
    }
};
